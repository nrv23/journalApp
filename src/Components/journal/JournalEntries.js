import React from 'react'
import { useSelector } from 'react-redux';
import { JournalEntry } from './JournalEntry';

export const JournalEntries = () => {

    const entries =  useSelector(state => state.notes.notes)
    return (
        <div className="journal__entries">
            {
                entries.map(entry => (<JournalEntry key={entry.id} {...entry} />))
            }
        </div>
    )
}
