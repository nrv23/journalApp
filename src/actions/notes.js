import { db } from '../firebase/firebaseConfig';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types';

export const starNewNote = () =>  {

    return async (dispatch,getState) => { // el segundo parÃ¡metro del callback que devuelve la funcion asincrona, es una funcion
        // que devuelve todo el state global actual
        try {
            
            const { auth:{uid}  } = getState();

            const newNote = {

                title: '',
                body: '',
                date: new Date().getTime()
            }

            const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
            dispatch(activeNote(docRef.id,newNote));
        } catch (error) {
            console.log(error)
        }
    }
}

export const activeNote = (id,note) => {

    return {
        type: types.NOTESACTIVE,
        payload: {
            ...note,
            id
        }
    }
}

export const startLoadingNotes = (uid) => {

    return async dispatch => { // redux-thunk permite la ejecucion de dispatchs asincronos
        const notes = await loadNotes(uid);

        dispatch(setNotes(notes));
    }
}

export const setNotes = notes => ({
    type: types.NOTESLOAD,
    payload: notes
})