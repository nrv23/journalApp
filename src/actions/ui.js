import { types } from "../types"



export const setError = (error) => {
    
    return {
        type: types.UISETERROR,
        payload: error
    }
}

export const unSetError = () => {
    
    return {
        type: types.UIUNSETERROR
    }
}

export const startLoading = () => {

    return {
        type: types.UISTARTLOADING,
        payload: true
    }
}


export const finishLoading = () => {

    return {
        type: types.UISTARTLOADING,
        payload: false
    }
}