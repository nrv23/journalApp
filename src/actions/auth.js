import  { firebase,googleAuthProvider } from '../firebase/firebaseConfig';
import { types } from '../types/index';
import { clearNotes } from './notes';
import { startLoading,finishLoading, setError,unSetError } from './ui';


export const startLoginEmailPassword = (email, pass) => {
    
    return dispatch => { // retornar un callback para ejecutar un action asincrono
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email,pass).then(({user:{uid,displayName}}) => {
            dispatch(login(uid,displayName))
            dispatch(finishLoading());
            dispatch(unSetError());
        })
        .catch(e => {
            dispatch(finishLoading());
            console.log(e.code)

            if(e.code === 'auth/invalid-email'|| e.code === 'auth/wrong-password') { 
                dispatch(setError('Credenciales incorrectos'))
            } else {

                dispatch(setError("Hubo un error en el login"));
            }

        })
    }
}

export const login = (uid, displayName) => {

    return {
        type: types.LOGIN,
        payload: {
            uid,
            displayName
        }
    }
}

export const startGoogleLogin = () => {


    return dispatch => {
        firebase.auth().signInWithPopup(googleAuthProvider)
        .then(({user:{uid,displayName}}) => {
            dispatch(login(uid,displayName))
        })
    }
}

export const startRegisterWithEmailPasswordName = (email,password,name) => {

    return dispatch => {

        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async ({user}) => {

                await user.updateProfile({displayName: name});
                dispatch(login(user.uid,user.displayName))
            })
            .catch(e => {
                console.log(e.code)
            })
    }

}

export const startLogout = () => {

    return async dispatch => {

        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(clearNotes());
    }
}

export const logout = () => {

    return {
        type: types.LOGOUT
    }
}