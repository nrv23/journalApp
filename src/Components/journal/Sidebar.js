import React from 'react'
import { JournalEntries } from './JournalEntries'
import { useDispatch,useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { starNewNote } from '../../actions/notes';

export const Sidebar = () => {

    const {name} = useSelector(state => state.auth);
    const dispatch = useDispatch()

    const handleLogout = () => {

        
        dispatch(startLogout())

    }

    const handleAddNew = () => {

        dispatch(starNewNote())
    }

    return (
        <aside className="journal__sidebar">
            <div className="journal__sidebar-navbar">
                <i className="far fa-moon"></i>
                <h3 className="mt-5"> <span>{name}</span> </h3>

                <button  className="btn"
                    onClick={handleLogout}
                >
                    Logout
                </button>
            </div>

            <div className="journal__new-entry"
             onClick={handleAddNew}
            >
                <i className="far fa-calendar-plus fa-5x"></i>
                <p className="mt-5">New Entrie</p>
            </div>

            <JournalEntries />
        </aside>
    )
}
