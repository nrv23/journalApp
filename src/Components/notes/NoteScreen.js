import React,{useEffect, useRef} from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useSelector,useDispatch } from 'react-redux';
import useForm from '../../hooks/useForm';
import { activeNote } from '../../actions/notes';

export const NoteScreen = () => {

    const { active:note } = useSelector(state => state.notes);
    const dispatch = useDispatch();
    const [formValues,handleInputChange,reset] = useForm(note);

    const activeId = useRef(note.id);// el useRef almacena variables mutables que no va redibujar el componente si la variable cambia de valor

    useEffect(() => {
        if(note.id !== activeId.current) { // si el id de la nota actual es diferente a la referencia actual de id nota, resetear el formulario
            reset(note)
            activeId.current = note.id; //actualizar la referencia del id de la nota actual
        }
    }, [note,reset])

    const { body,title } = formValues;

    //actualizar el state de la nota activa

    useEffect(() => {
        dispatch(activeNote(formValues.id,{...formValues}));
    },[formValues,dispatch])

    return (
        <div className="notes__main-content">
            
        <NotesAppBar />

        <div className="notes__content">

            <input 
                type="text"
                name="title"
                placeholder="Some awesome title"
                className="notes__title-input"
                autoComplete="off"
                value={title}
                onChange={handleInputChange}
            />

            <textarea
                name="body"
                placeholder="What happened today"
                className="notes__textarea"
                value={body}
                onChange={handleInputChange}
            ></textarea>

            {
                note.url && 
                (<div className="notes__image">
                    <img 
                        src={`${note.url}`}
                        alt="imagen"
                    />
                </div>)
            }


        </div>

    </div>
    )
}
