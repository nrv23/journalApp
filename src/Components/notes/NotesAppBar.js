import React from 'react'
import { useSelector,useDispatch } from 'react-redux'
import { startSaveNote } from '../../actions/notes';

export const NotesAppBar = () => {

    const { active:note } = useSelector(state => state.notes)
    const dispatch = useDispatch();

    const handleUpdate = () => {

        dispatch(startSaveNote(note));
    }
    return (
        <div className="notes__appbar">
            <span>28 de agosto 2020</span>

            <div>
                <button className="btn">
                    Picture
                </button>

                <button className="btn" onClick={handleUpdate}>
                    Save
                </button>
            </div>
        </div>
    )
}
