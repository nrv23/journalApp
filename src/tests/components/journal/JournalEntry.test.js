
import { mount } from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store' //ES6 modules
//const { configureStore } = require('redux-mock-store') //CommonJS
import thunk from 'redux-thunk'; // middleware para acciones asincronas
import { activeNote } from '../../../actions/notes';
import { JournalEntry } from '../../../Components/journal/JournalEntry';

const middlewares = [thunk]
const mockStore = configureStore(middlewares);
const initState = {}
let store = mockStore(initState);
store.dispatch= jest.fn();

const newNote= {
    id: "12345",
    title:"hola",
    body:"mundo",
    date: Date.now()
}

const wrapper = mount( // cuando da el error del router, usa el memoryRouter para simular una ruta
    <Provider store= {store}>
        <JournalEntry {...newNote}/>
    </Provider>
);

describe('Prueba es <JournalEntry />', () => {
    
    test('Debe mostrarse correctamente ', () => {
         expect(wrapper).toMatchSnapshot()
    })
    

    test('Debe activar la nota', () => {
        
        wrapper.find('.journal__entry').simulate('click');


        expect(store.dispatch).toHaveBeenCalledWith(activeNote(newNote.id,{...newNote}));
    })
    
})
