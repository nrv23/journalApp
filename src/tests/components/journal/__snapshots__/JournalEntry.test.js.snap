// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba es <JournalEntry /> Debe mostrarse correctamente  1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      hola
    </p>
    <p
      className="journal__entry-content"
    >
      mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Tuesday
    </span>
    <h4>
      16th
    </h4>
  </div>
</div>
`;
